### PR TITLE
Desk

### PR DESCRIPTION
**Pull request by teodoratsaneva**

### CODE REVIEW

File: DeskReserve/Controllers/DesksController.cs
Comment (MartinPetrakiev) [inline DeskReserve/Controllers/DesksController.cs:1]: The logic in each controller endpoint is simple for now, but it will be better practice to sperate this logic in a Service.

File: DeskReserve/Data/DBContext/Entity/Desk.cs
Comment (MartinPetrakiev) [inline DeskReserve/Data/DBContext/Entity/Desk.cs:1]: Try including more Data Annotations to the Entity. For example which properties should be required in the DB, IsOccupied default value on creation?

File: DeskReserve/Controllers/DesksController.cs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+using DeskReserve.Data.DBContext;
+using DeskReserve.Data.DBContext.Entity;
+
+namespace DeskReserve.Controllers
+{
+    [Route("api/[controller]")]
+    [ApiController]
+    public class DesksController : ControllerBase
+    {
+        private readonly ApplicationDbContext _context;
```
Comment (Dolen-Chalgar) [inline DeskReserve/Controllers/DesksController.cs]: Why controller needs to know about dbcontext

File: DeskReserve/Controllers/DesksController.cs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+        public DesksController(ApplicationDbContext context)
+        {
+            _context = context;
```
Comment (Dolen-Chalgar) [inline DeskReserve/Controllers/DesksController.cs]: Null check is needed here

File: DeskReserve/Controllers/DesksController.cs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        }
+
+        // GET: api/Desks
```
Comment (Dolen-Chalgar) [inline DeskReserve/Controllers/DesksController.cs]: Remove unnecessary comment HTTP attribute is explicit enough

File: DeskReserve/Controllers/DesksController.cs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        public async Task<ActionResult<Desk>> GetDesk(Guid id)
+        {
+            var desk = await _context.Desks.FindAsync(id);
+
+            if (desk == null)
+            {
+                return NotFound();
+            }
+
+            return desk;
```
Comment (Dolen-Chalgar) [inline DeskReserve/Controllers/DesksController.cs]: Be consistent withy the return types of the controller. Either use action result or return the entity itself

File: DeskReserve/Controllers/DesksController.cs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        // GET: api/Desks
+        [HttpGet]
+        public async Task<ActionResult<IEnumerable<Desk>>> GetDesk()
+        {
+            return await _context.Desks.ToListAsync();
+        }
+
+        // GET: api/Desks/5
+        [HttpGet("{id}")]
+        public async Task<ActionResult<Desk>> GetDesk(Guid id)
```
Comment (Dolen-Chalgar) [inline DeskReserve/Controllers/DesksController.cs]: Should the controller care how the data is handled?

File: DeskReserve/Controllers/DesksController.cs [hunk 6]
```diff
@@ -0,0 +1,108 @@
+﻿using System;
```
Comment (Dolen-Chalgar) [inline DeskReserve/Controllers/DesksController.cs]: Please add unit test for ALL the logic you have written

File: DeskReserve/Controllers/DesksController.cs [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+                {
+                    return NotFound();
+                }
+                else
+                {
+                    throw;
+                }
+            }
+
+            return NoContent();
```
Comment (Dolen-Chalgar) [inline DeskReserve/Controllers/DesksController.cs]: This logic should be isolated in the service

File: DeskReserve/Controllers/DesksController.cs [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        // To protect from overposting attacks, see https://go.microsoft.com/fwlink/?linkid=2123754
+        [HttpPut("{id}")]
+        public async Task<IActionResult> PutDesk(Guid id, Desk desk)
+        {
+            if (id != desk.DeskId)
+            {
+                return BadRequest();
+            }
+
+            _context.Entry(desk).State = EntityState.Modified;
```
Comment (Dolen-Chalgar) [inline DeskReserve/Controllers/DesksController.cs]: Do we really need to explicitly check the state of the object in the DB context to know it is modified?

File: DeskReserve/Controllers/DesksController.cs [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+            try
+            {
+                await _context.SaveChangesAsync();
+            }
+            catch (DbUpdateConcurrencyException)
+            {
+                if (!DeskExists(id))
```
Comment (Dolen-Chalgar) [inline DeskReserve/Controllers/DesksController.cs]: Why catch exception and not handle it?

File: DeskReserve/Controllers/DesksController.cs [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+                {
+                    return NotFound();
+                }
+                else
+                {
+                    throw;
```
Comment (Dolen-Chalgar) [inline DeskReserve/Controllers/DesksController.cs]: ?

File: DeskReserve/Controllers/DesksController.cs [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+
+        // POST: api/Desks
+        // To protect from overposting attacks, see https://go.microsoft.com/fwlink/?linkid=2123754
+        [HttpPost]
+        public async Task<ActionResult<Desk>> PostDesk(Desk desk)
+        {
+            _context.Desks.Add(desk);
+            await _context.SaveChangesAsync();
+
+            return CreatedAtAction("GetDesk", new { id = desk.DeskId }, desk);
```
Comment (Dolen-Chalgar) [inline DeskReserve/Controllers/DesksController.cs]: Wrorg Action param. Lookup this:https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.controllerbase.createdataction?view=aspnetcore-7.0

File: DeskReserve/Controllers/DesksController.cs (hunk 5 above)
Comment (Dolen-Chalgar) [inline DeskReserve/Controllers/DesksController.cs]: Why Get Desk when this DesksController. Just Get is good enough

File: DeskReserve/Controllers/DesksController.cs [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+            return await _context.Desks.ToListAsync();
+        }
+
+        // GET: api/Desks/5
+        [HttpGet("{id}")]
+        public async Task<ActionResult<Desk>> GetDesk(Guid id)
+        {
+            var desk = await _context.Desks.FindAsync(id);
+
+            if (desk == null)
```
Comment (Dolen-Chalgar) [inline DeskReserve/Controllers/DesksController.cs]: use reference check here instead of value

File: DeskReserve/Controllers/DesksController.cs [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+                return NotFound();
+            }
+
+            return desk;
+        }
+
+        // PUT: api/Desks/5
+        // To protect from overposting attacks, see https://go.microsoft.com/fwlink/?linkid=2123754
+        [HttpPut("{id}")]
+        public async Task<IActionResult> PutDesk(Guid id, Desk desk)
```
Comment (Dolen-Chalgar) [inline DeskReserve/Controllers/DesksController.cs]: Use request model instead of Guid and Desk

File: DeskReserve/Controllers/DesksController.cs [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+        {
```
Comment (Dolen-Chalgar) [inline DeskReserve/Controllers/DesksController.cs]: Reconsider solution. Do we really need all these validations

File: DeskReserve/Controllers/DesksController.cs [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+                return NotFound();
+            }
+
+            _context.Desks.Remove(desk);
+            await _context.SaveChangesAsync();
+
+            return NoContent();
+        }
+
+        private bool DeskExists(Guid id)
```
Comment (Dolen-Chalgar) [inline DeskReserve/Controllers/DesksController.cs]: Why do we have business logic on a controller level

File: DeskReserve/Controllers/DesksController.cs [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+        }
+
+        // DELETE: api/Desks/5
+        [HttpDelete("{id}")]
+        public async Task<IActionResult> DeleteDesk(Guid id)
+        {
+            var desk = await _context.Desks.FindAsync(id);
+            if (desk == null)
+            {
+                return NotFound();
```
Comment (Dolen-Chalgar) [inline DeskReserve/Controllers/DesksController.cs]: Multiple return point in one method

File: DeskReserve/Data/DBContext/Entity/Desk.cs [hunk 17]
```diff
@@ -0,0 +1,15 @@
+using System.ComponentModel.DataAnnotations;
+using System.ComponentModel.DataAnnotations.Schema;
+
+namespace DeskReserve.Data.DBContext.Entity
+{
+	[Table("desk")]
+	public class Desk
+	{
+		[Key]
+		public Guid DeskId { get; set; }
```
Comment (Dolen-Chalgar) [inline DeskReserve/Data/DBContext/Entity/Desk.cs:11]: Can we establish the relation via entity framework's tools

File: DeskReserve/Domain/DeskService.cs [hunk 18]
```diff
@@ -0,0 +1,66 @@
+using DeskReserve.Data.DBContext.Entity;
+using DeskReserve.Data.DBContext;
+using Microsoft.EntityFrameworkCore;
+
+namespace DeskReserve.Domain
+{
+	public class DeskService
```
Comment (Dolen-Chalgar) [inline DeskReserve/Domain/DeskService.cs]: What about using an interface :)

File: DeskReserve/Domain/DeskService.cs [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+	{
+		private readonly ApplicationDbContext _dbContext;
+
+		public DeskService(ApplicationDbContext dbContext)
```
Comment (Dolen-Chalgar) [inline DeskReserve/Domain/DeskService.cs]: Register the service and use the IOC instance

File: DeskReserve/Domain/DeskService.cs [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+		{
+			_dbContext = dbContext ?? throw new ArgumentNullException(nameof(dbContext));
+		}
+
+		public List<Desk> GetAllDesks()
```
Comment (Dolen-Chalgar) [inline DeskReserve/Domain/DeskService.cs]: again Desk in function names is not needed

Comment (Dolen-Chalgar) [inline DeskReserve/Domain/DeskService.cs]: There is more appropriate interface here

File: DeskReserve/Domain/DeskService.cs [hunk 21]
```diff
@@ -0,0 +1,66 @@
+using DeskReserve.Data.DBContext.Entity;
+using DeskReserve.Data.DBContext;
+using Microsoft.EntityFrameworkCore;
+
+namespace DeskReserve.Domain
```
Comment (Dolen-Chalgar) [inline DeskReserve/Domain/DeskService.cs]: Why do we have duplicating logic in controller and service?

File: DeskReserve/Domain/DeskService.cs [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+		{
+			return _dbContext.Desks.ToList();
+		}
+
+		public async Task<Desk> GetDesk(Guid id)
+		{
+			return await _dbContext.Desks.FindAsync(id);
+		}
+
+		public async Task<bool> UpdateDesk(Guid id, Desk desk)
```
Comment (Dolen-Chalgar) [inline DeskReserve/Domain/DeskService.cs]: Again single point of return

File: DeskReserve/Domain/DeskService.cs [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+			}
+
+			_dbContext.Entry(desk).State = EntityState.Modified;
+
+			try
+			{
+				await _dbContext.SaveChangesAsync();
+				return true;
+			}
+			catch (DbUpdateConcurrencyException)
```
Comment (Dolen-Chalgar) [inline DeskReserve/Domain/DeskService.cs]: Does this mean catching concurrency exception is the same as passing invalid id since in both places we return false

File: DeskReserve/Domain/DeskService.cs [hunk 24]
```diff
@@ -0,0 +1,66 @@
...
+			{
+				return false;
+			}
+
+			_dbContext.Desks.Remove(desk);
+			await _dbContext.SaveChangesAsync();
+			return true;
+		}
+	}
+}
```
Comment (Dolen-Chalgar) [inline DeskReserve/Domain/DeskService.cs]: Add end of lines in file

File: DeskReserve/Program.cs [hunk 25]
```diff
@@ -27,6 +27,8 @@
 
 builder.Services.AddScoped<DogService, DogService>();
 builder.Services.AddScoped<DogController, DogController>();
+builder.Services.AddScoped<DeskService, DeskService>();
```
Comment (Dolen-Chalgar) [inline DeskReserve/Program.cs]: When you add the interfaces use them isntead

File: DeskReserve/appsettings.json [hunk 26]
```diff
@@ -6,7 +6,7 @@
     }
   },
   "ConnectionStrings": {
-    "DefaultConnection": "Server=myServerAddress;Database=myDataBase;User Id=myUsername;Password=myPassword;"
+    "DefaultConnection": "Server=MihailPC;Database=test123;Trusted_Connection=True"
```
Comment (Dolen-Chalgar) [inline DeskReserve/appsettings.json:9]: Why you commit your connection string

File: Initial.sql [hunk 27]
```diff
@@ -8,5 +8,13 @@ CREATE TABLE [Dog]
 	[name] NVARCHAR(255) 
 )
 
+CREATE TABLE [Desk](
```
Comment (Dolen-Chalgar) [inline Initial.sql]: PLEASE create new script for each DDL or DML operaion

**Review by Dolen-Chalgar — CHANGES_REQUESTED** [review]

File: DeskReserve/Data/DBContext/ApplicationDbContext .cs [hunk 28]
```diff
@@ -10,5 +10,6 @@ public ApplicationDbContext(DbContextOptions<ApplicationDbContext> options)
 
         public DbSet<Dog> Dogs { get; set; }
```
Comment (Tanya-Zheleva) [inline DeskReserve/Data/DBContext/ApplicationDbContext .cs]: Why is this here? Always check the code before creating a PR

File: DeskReserve/Domain/DeskService.cs [hunk 29]
```diff
@@ -0,0 +1,66 @@
...
+		public async Task<bool> DeleteDesk(Guid id)
+		{
+			var desk = await _dbContext.Desks.FindAsync(id);
+			if (desk == null)
+			{
+				return false;
+			}
+
+			_dbContext.Desks.Remove(desk);
+			await _dbContext.SaveChangesAsync();
```
Comment (Tanya-Zheleva) [inline DeskReserve/Domain/DeskService.cs]: Fix formatting & add a new line before the return statement

File: Initial.sql [hunk 30]
```diff
@@ -8,5 +8,13 @@ CREATE TABLE [Dog]
...
+	[DeskId] UNIQUEIDENTIFIER PRIMARY KEY,
+	[DeskNumber] INT,
+	[RoomId] UNIQUEIDENTIFIER,
+	[IsOccupied] BIT,
+	[IsStatic] BIT
+)
+
 INSERT INTO Dog(id, age , name)
```
Comment (Tanya-Zheleva) [inline Initial.sql]: Remove the old scripts, they are not relevant to this project

File: DeskReserve/Controllers/DesksController.cs (hunk 1 above)
Comment (Tanya-Zheleva) [inline DeskReserve/Controllers/DesksController.cs]: Is there any specific reason why you are using an underscore in the beginning of the name?

**Review by Tanya-Zheleva — CHANGES_REQUESTED** [review]

File: Initial.sql [hunk 31]
```diff
@@ -1,12 +1,10 @@
 CREATE DATABASE test123
 USE test123
 
-CREATE TABLE [Dog]
-(
-	[id] UNIQUEIDENTIFIER PRIMARY KEY,
-	[age] INT,
-	[name] NVARCHAR(255) 
+CREATE TABLE [Desk](
+	[DeskId] UNIQUEIDENTIFIER PRIMARY KEY,
```
Comment (Dolen-Chalgar) [inline Initial.sql]: Again add this table creation on a new script

File: DeskReserve/Domain/DeskService.cs [hunk 32]
```diff
@@ -1,66 +1,57 @@
...
+using DeskReserve.Mapper;
+using DeskReserve.Exception;
 
 namespace DeskReserve.Domain
 {
-	public class DeskService
+	public class DeskService : IDeskService
 	{
-		private readonly ApplicationDbContext _dbContext;
+		private readonly IDeskRepository repository;
```
Comment (Dolen-Chalgar) [inline DeskReserve/Domain/DeskService.cs]: The convention is that private fields should start with an underscore

File: DeskReserve/Domain/DeskService.cs [hunk 33]
```diff
@@ -1,66 +1,57 @@
...
 		}
 
-		public async Task<bool> UpdateDesk(Guid id, Desk desk)
+		public async Task<bool> UpdateOneAsync(Guid id, DeskDto deskDto)
 		{
-			if (id != desk.DeskId)
-			{
-				return false;
-			}
+			var desk = await repository.GetById(id) ?? throw new EntityNotFoundException();
```
Comment (Dolen-Chalgar) [inline DeskReserve/Domain/DeskService.cs]: Why we are not using the extension methods here?

File: DeskReserve/Domain/DeskService.cs [hunk 34]
```diff
@@ -1,66 +1,57 @@
...
-			catch (DbUpdateConcurrencyException)
-			{
-				return false;
-			}
+			return await repository.Update(desk);
 		}
 
-		public async Task<Desk> CreateDesk(Desk desk)
+
+		public async Task<bool> CreateOneAsync(DeskDto deskDto)
```
Comment (Dolen-Chalgar) [inline DeskReserve/Domain/DeskService.cs]: Seeing the param is a single element CreateOne for example is good enough

File: DeskReserve/Exceptions/EntityNotFoundException.cs [hunk 35]
```diff
@@ -0,0 +1,24 @@
+﻿using System.Runtime.Serialization;
+
+namespace DeskReserve.Exception
+{
+    [Serializable]
```
Comment (Dolen-Chalgar) [inline DeskReserve/Exceptions/EntityNotFoundException.cs]: Why exception is serialzable?

File: DeskReserve/Domain/DeskService.cs [hunk 36]
```diff
@@ -1,66 +1,57 @@
 using DeskReserve.Data.DBContext.Entity;
```
Comment (Dolen-Chalgar) [inline DeskReserve/Domain/DeskService.cs]: Move servcices to a Service folder

File: DeskReserve/Repository/DeskRepository.cs [hunk 37]
```diff
@@ -0,0 +1,55 @@
...
+		private readonly ApplicationDbContext _context;
+
+		public DeskRepository(ApplicationDbContext context)
+		{
+			_context = context ?? throw new ArgumentNullException(nameof(context));
+		}
+
+		public async Task<IEnumerable<Desk>> GetAll()
+		{
+			return await _context.Desks.ToListAsync() ?? throw new ArgumentNullException();
```
Comment (Dolen-Chalgar) [inline DeskReserve/Repository/DeskRepository.cs]: Why are we using ArgumentNullException?

File: DeskReserve/Repository/DeskRepository.cs [hunk 38]
```diff
@@ -0,0 +1,55 @@
...
+		public async Task<Desk> GetById(Guid id)
+		{
+			return await _context.Desks.FindAsync(id) ?? throw new EntityNotFoundException();
+		}
+
+		public async Task<bool> Update(Desk desk)
+		{
+			var existingDesk = await _context.Desks.FindAsync(desk.DeskId) ?? throw new EntityNotFoundException();
+
+			existingDesk.DeskNumber = desk.DeskNumber;
```
Comment (Dolen-Chalgar) [inline DeskReserve/Repository/DeskRepository.cs]: We should not map on two places

File: DeskReserve_Tests/DeskReserve_Tests - Backup.csproj [hunk 39]
```diff
@@ -0,0 +1,25 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (Dolen-Chalgar) [inline DeskReserve_Tests/DeskReserve_Tests - Backup.csproj]: Why do you have a backup .sln file :D

File: DeskReserve_Tests/DeskReserve_Tests.sln [hunk 40]
```diff
@@ -0,0 +1,25 @@
+﻿
+Microsoft Visual Studio Solution File, Format Version 12.00
+# Visual Studio Version 17
```
Comment (Dolen-Chalgar) [inline DeskReserve_Tests/DeskReserve_Tests.sln:3]: There is something going on here with these multiple .sln files. We should have only one

File: DeskReserve_Tests/DeskServiceTest.cs [hunk 41]
```diff
@@ -0,0 +1,111 @@
...
+
+namespace Desk_Reserve.Tests
+{
+	[TestFixture]
+	public class DeskServiceTests
+	{
+		[Test]
+		public async Task GetAllAsync_ReturnsAllDesks()
+		{
+			var repositoryMock = new Mock<IDeskRepository>();
```
Comment (Dolen-Chalgar) [inline DeskReserve_Tests/DeskServiceTest.cs]: Since we are using the repository for all test try to isolate this in a method that mock our dependencies. Look up TestFixture

File: DeskReserve_Tests/DeskServiceTest.cs [hunk 42]
```diff
@@ -0,0 +1,111 @@
+﻿using DeskReserve.Data.DBContext.Entity;
```
Comment (Dolen-Chalgar) [inline DeskReserve_Tests/DeskServiceTest.cs:1]: Add unit tests for the mapper

**Review by Dolen-Chalgar — CHANGES_REQUESTED** [review]

File: DeskReserve/Exceptions/EntityNotFoundException.cs (hunk 35 above)
Comment (teodoratsaneva) [reply DeskReserve/Exceptions/EntityNotFoundException.cs]: Serializing exceptions can be beneficial when you have errors occurring in different components of your system. If these components are distributed across different locations, serialization can help easily share error information between them.   For example, if an error is thrown in the repository and you want to pass it back to the controller through your service, serialization can facilitate this process.

File: DeskReserve/Controllers/DesksController.cs (hunk 1 above)
Comment (teodoratsaneva) [reply DeskReserve/Controllers/DesksController.cs]: Private fields should start with an underscore

**Review by Dolen-Chalgar — APPROVED** [review]

**Review by Dolen-Chalgar — APPROVED** [review]
